### DISCUSSION TITLE
Manipulating and Analyzing Output Values of  Vectors

### DISCUSSION THREAD

**Discussion by sSajjad90** [idaholab/moose#24772] [Q&A General]
I am currently working on developing a code based on the material class to calculate the number density of clusters in three phases. In my code, I have three vectors of materials, and their sizes remain constant throughout the iterations. I have been exploring options to manipulate the values of these vectors using Vectorpostprocessor (VPP) or reporters, or any other alternative method available. However, it seems that the existing VPPs and reporters are designed for real values and not specifically for vector materials.

I have attempted to adapt SamplerBase VPPs, such as LineMaterialRealSampler and LineValueSampler, and I have also tried creating a new code based on the existing base by utilizing references, objects, and pointers to access the vector values on the material-based code. Unfortunately, I have encountered errors throughout these attempts.

Now, I am seeking suggestions or assistance on how to effectively manipulate the values of these vectors. It's important to note that only real vector values need to be manipulated, and the `qp` is not considered in this context.
Below, you can find a portion of the `Unified.C `code and an example `input file` for reference.

```
[Material]
...

[./Unified_Calculation]
    type = Unified
    ...
    ...
    Rho_calc_T = 'Rho_calc_T'   //---> This is a vector of material, with a constant size (size=3000)
    Rho_calc_TP = 'Rho_calc_TP'   //---> This is a vector of material, with a constant size (size=3000)
    Rho_calc_TDP = 'Rho_calc_TDP'  //---> This is a vector of material, with a constant size (size=3000)
  [../]
[]
```

```
....
a1_calc_TDP[_qp].resize(MaxMatSize);  //--> it has a constant size for all iteration, MaxMatSize = 3000
rho_calc_TDP[_qp].assign(MaxMatSize, 0.0);
....
for (size_t i = 0; i < Interpol_Size; i++) {
    Alpha_Vals[phi3][i] = Alpha_Vector[phi3][i];
    Beta_Vals[phi3][i] = Beta_Vector[phi3][i];
    Rho_Old_Vals[phi3][i] = (*rho_calc_TDP_old)[_qp][i];
  }

  Thomas_Algorithm_Func(Alpha_Vals,Beta_Vals,Rho_Old_Vals,Del_Check,phi3);

  for (size_t i = 2; i < Phase_Num; i++) {
    for (size_t j = 0; j < (*Matrix_Size_prop_old)[_qp]; j++) {        //--> Max ( (*Matrix_Size_prop_old)[_qp]= 3000)
      rho_calc_TDP[_qp][j] = Rho_out_Vector[phi3][j];
....
```

**Comment by GiudGiud** [comment]
hello

So first, for context, are you using a single simulation to compute 3000 clusters?
Could these simulations be separated numerically using multiapps?

Secondly, are you working with vectors of material properties? Or a vector (really array, vector is size 3 only) material property?
It looks to me like the latter.

Finally,
this doesnt look right
```
    for (size_t j = 0; j < (*Matrix_Size_prop_old)[_qp]; j++) {        //--> Max ( (*Matrix_Size_prop_old)[_qp]= 3000)
```
could it be
```
    for (size_t j = 0; j < (*Matrix_Size_prop_old)[_qp].size(); j++) {        //--> Max ( (*Matrix_Size_prop_old)[_qp]= 3000)
```
?
Guillaume

**Reply by sSajjad90** [reply]
Thank you for your consideration.

Regarding the vectors, they represent the number densities of various sizes of different types of precipitates. (For example,` j = 50` corresponds to the number density of the precipitate `phi3` phase with a specific size of `5 nm,` stored in `Rho_out_Vector[phi3][j]`,  or` j = 500` represents the number density of precipitates with other sizes, such as `93 nm`.) These vectors are obtained for tracking the evolution of number densities for different sizes of precipitates over time in a transient simulation.

Regarding the use of the multiapp, if there is a pre-existing transfer code designed for transferring these vectors within the simulation, it would be suitable for my needs (specifically for the next steps of my project).
Since you mentioned that only vectors can be used, I employed other material properties calculated in the input file to compute these vectors, but these vectors of material properties are utilized to return results to the input file; there are no other computations involved in this simulation.

Thank you once again for your assistance.
Sajjad

**Reply by GiudGiud** [reply]
we do not have code to transfer array material properties, or maps right now.
However it s not that difficult to write new transfers that are only changing the source of the data, as long as the target (a variable) is the same. 

so Rho_out_Vector[phi3][j] has two indexes, is the underlying data type a vector of vector?

**Reply by sSajjad90** [reply]
Yes. it is a vector of vector, but it would be possible to change it into three separate vectors instead,
like: `Phi1_Rho_out_Vector[i]`, `Phi2_Rho_out_Vector[i]`, and `Phi3_Rho_out_Vector[i]`

**Reply by GiudGiud** [reply]
That might be easier, there's a tad more capability with array material properties than with arbitrary tensors/matrix.
Still no transfer through.

**Reply by sSajjad90** [reply]
I'm still pondering how to tackle this issue. Any suggestions would be greatly appreciated.

**Reply by GiudGiud** [reply]
which issue in particular? the transfer/

**Reply by sSajjad90** [reply]
Sending a vector of real values to the vectorpostprocessor or transfer it through a multiapp.

Thanks in advance for your consideration,
Sajjad

**Reply by GiudGiud** [reply]
you cant send to a vectorpostprocessor.

You can use a VectorPostprocessor to collect data from the simulation.
Here's the full list of the ones you can use:
https://mooseframework.inl.gov/syntax/VectorPostprocessors/index.html

But here the vectors are different at each location in space right? It s like a material property?
